cmd/containerd-shim-runc-v2/task/service.go

```diff
@@ -168,33 +171,58 @@ func (s *service) preStart(c *runc.Container) (handleStarted func(*runc.Containe
 		}
 	}
 
-	handleStarted = func(c *runc.Container, p process.Process, muLocked bool) {
+	handleStarted = func(c *runc.Container, p process.Process) {
 		var pid int
 		if p != nil {
 			pid = p.Pid()
 		}
 
+		var init bool
+		if _, ok := p.(*process.Init); ok {
+			init = true
+		}
 		s.lifecycleMu.Lock()
+
+		var initExits []runcC.Exit
+		var initCp *containerProcess
+		var shouldProcessInitExit bool
+		if !init {
+			initPid := c.Pid()
+			iExits, initExited := exits[initPid]
+			initExits = iExits
+			for _, possibleCps := range s.running[initPid] {
+				if possibleCps.Container == c {
+					possibleCps := possibleCps
+					initCp = &possibleCps
+				}
+			}
+			*initCp.PendingExecCounter--
+			if initExited && *initCp.PendingExecCounter == 0 {
+				shouldProcessInitExit = true
+				delete(s.running, initPid)
+			}
+		}
+
 		ees, exited := exits[pid]
 		delete(s.exitSubscribers, &exits)
 		exits = nil
-		if pid == 0 { // no-op
-			s.lifecycleMu.Unlock()
-		} else if exited {
+		if pid == 0 || exited {
 			s.lifecycleMu.Unlock()
 			for _, ee := range ees {
-				if muLocked {
-					s.handleProcessExit(ee, c, p)
-				} else {
-					s.mu.Lock()
-					s.handleProcessExit(ee, c, p)
-					s.mu.Unlock()
+				s.handleProcessExit(ee, c, p)
+			}
+			if shouldProcessInitExit {
+				for _, eee := range initExits {
+					s.handleProcessExit(eee, initCp.Container, initCp.Process)
 				}
```

Comment:
Without pidfd, it's impossible to handle exits correctly if PIDs get recycled. Although we can't handle them correctly, we can at least try to handle them in a not-too-broken way. If control reaches here, the entire slice of `s.running[initPid]` has been deleted from the map. Only the exit for the `initCp` process gets broadcast. All the other "running" processes with the same PID are silently dropped. Without the exit events, a higher-level runtime might be stuck forever with leaked "zombie" state corresponding to those dropped processes. An exit event needs to be broadcast for each container process removed from `s.running` as we can't broadcast the events after we have discarded the information needed to populate them.

Reply:
Fixed!